ruma-federation-api/src/transactions/send_transaction_message/edu.rs

```diff
@@ -0,0 +1,288 @@
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct ReceiptMap {
+    /// Read receipts for users in the room.
+    #[serde(rename = "m.read")]
+    pub read: BTreeMap<UserId, ReceiptData>,
+}
+
+/// Metadata about the event that was last read and when.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct ReceiptData {
+    /// Metadata for the read receipt.
+    pub data: Receipt,
+
+    /// The extremity event ID the user has read up to.
+    pub event_ids: Vec<EventId>,
+}
+
+/// The content for "m.typing" Edu.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct TypingContent {
+    /// The typing notification.
+    pub notifi: TypingNotice,
+}
+
+/// Notification of a user typing.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct TypingNotice {
+    /// The room where the user's typing status has been updated.
+    pub room_id: RoomId,
+
+    /// The user ID that has had their typing status changed.
+    pub user_id: UserId,
+
+    /// Whether the user is typing in the room or not.
+    pub typing: bool,
+}
+
+/// The description of the direct-to- device message.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct DeviceListUpdateContent {
+    /// The user ID who owns the device.
+    pub user_id: UserId,
+
+    /// The ID of the device whose details are changing.
+    pub device_id: DeviceIdBox,
+
+    /// The public human-readable name of this device. Will be absent if the device has no name.
+    pub device_display_name: String,
+
+    /// An ID sent by the server for this update, unique for a given user_id.
+    pub stream_id: UInt,
+
+    /// The stream_ids of any prior m.device_list_update EDUs sent for this user
+    // which have not been referred to already in an EDU's prev_id field.
+    #[serde(default, skip_serializing_if = "Vec::is_empty")]
+    pub prev_id: Vec<UInt>,
+
+    /// True if the server is announcing that this device has been deleted.
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub deleted: Option<bool>,
+
+    /// The updated identity keys (if any) for this device.
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub keys: Option<DeviceKeys>,
+}
+
+/// The description of the direct-to- device message.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct DirectDeviceContent {
+    /// The user ID of the sender.
+    pub sender: UserId,
+
+    /// Event type for the message.
+    #[serde(rename = "type")]
+    pub ev_type: String,
+
+    /// Unique utf8 string ID for the message, used for idempotence.
+    pub message_id: String,
+
+    // TODO: https://matrix.org/docs/spec/server_server/r0.1.4#m-direct-to-device-schema
+    /// The contents of the messages to be sent. These are arranged in a map
+    /// of user IDs to a map of device IDs to message bodies. The device ID may
+    /// also be *, meaning all known devices for the user.
+    pub messages: BTreeMap<String, BTreeMap<DeviceIdBox, JsonValue>>,
```

Comment:
Does anyone know which type the `JsonValue` should be here I don't know a whole lot about the e2e stuff.

@timokoesters

Reply:
From the spec example looks like `RoomKeyEventContent`. Also `String` should be `UserId` instead AFAICT.